test/functional/service_controller_test.go

```diff
@@ -0,0 +1,145 @@
+/*
+Copyright 2022.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package functional_test
+
+import (
+	"fmt"
+	"time"
+
+	"github.com/google/uuid"
+	. "github.com/onsi/ginkgo/v2"
+	. "github.com/onsi/gomega"
+
+	"k8s.io/apimachinery/pkg/types"
+)
+
+var _ = Describe("Service controller", func() {
+
+	var serviceName types.NamespacedName
+	var dnsMasqName types.NamespacedName
+	var svcDNSData types.NamespacedName
+	var namespace string
+
+	BeforeEach(func() {
+		// NOTE(gibi): We need to create a unique namespace for each test run
+		// as namespaces cannot be deleted in a locally running envtest. See
+		// https://book.kubebuilder.io/reference/envtest.html#namespace-usage-limitation
+		namespace = uuid.New().String()
+		th.CreateNamespace(namespace)
+		// We still request the delete of the Namespace to properly cleanup if
+		// we run the test in an existing cluster.
+		DeferCleanup(th.DeleteNamespace, namespace)
+
+	})
+
+	When("A Service is created with dnsmasq annotation", func() {
+		BeforeEach(func() {
+			instance := CreateDNSMasq(namespace, GetDefaultDNSMasqSpec())
+			dnsMasqName = types.NamespacedName{
+				Name:      instance.GetName(),
+				Namespace: namespace,
+			}
+			serviceName = types.NamespacedName{
+				Name:      "some-service",
+				Namespace: namespace,
+			}
+			svcDNSData = types.NamespacedName{
+				Name:      fmt.Sprintf("%s-svc", dnsMasqName.Name),
+				Namespace: namespace,
+			}
+
+			svc := CreateLoadBalancerService(serviceName, true)
+
+			DeferCleanup(th.DeleteInstance, svc)
+			DeferCleanup(th.DeleteInstance, instance)
+		})
+
+		It("should have created a DNSData with hosts data", func() {
+			svc := GetService(serviceName)
+			Expect(svc).To(Not(BeNil()))
+
+			// seems a wait is required, otherwise it can fail?!?
+			time.Sleep(1000 * time.Millisecond)
```

Comment:
OK, I could reproduce an intermittent failure locally if I remove this sleep.

I needed to add couple of more log statement to get a clear view what is the sequence of events.

Successful sequence:
```
  2023-05-11T10:48:39.834+0200	INFO	---Test---	Creating	{"raw": {"apiVersion":"network.openstack.org/v1beta1","kind":"DNSMasq","metadata":{"name":"bbc8d956-06c9-47f4-bd0e-f14a68c65863","namespace":"40218426-5c50-4f36-9a1b-58583d72ee5f"},"spec":{"containerImage":"test-dnsmasq-container-image","dnsDataLabelSelectorValue":"dnsdata","externalEndpoints":[{"ipAddressPool":"ctlplane","loadBalancerIPs":["internal-lb-ip-1","internal-lb-ip-2"]}],"options":[{"key":"server","values":["1.1.1.1"]}]}}}
  2023-05-11T10:48:39.934+0200	INFO	Starting workers	{"controller": "service", "controllerGroup": "", "controllerKind": "Service", "worker count": 1}
  2023-05-11T10:48:39.934+0200	INFO	Starting workers	{"controller": "dnsdata", "controllerGroup": "network.openstack.org", "controllerKind": "DNSData", "worker count": 1}
  2023-05-11T10:48:39.934+0200	INFO	Starting workers	{"controller": "dnsmasq", "controllerGroup": "network.openstack.org", "controllerKind": "DNSMasq", "worker count": 1}
  2023-05-11T10:48:39.934+0200	INFO	Reconciling	{"controller": "service", "controllerGroup": "", "controllerKind": "Service", "Service": {"name":"kubernetes","namespace":"default"}, "namespace": "default", "name": "kubernetes", "reconcileID": "48164718-f7d5-4c47-987d-8039739dc7c7", "req": "default/kubernetes"}
  2023-05-11T10:48:41.860+0200	INFO	Reconciling	{"controller": "dnsmasq", "controllerGroup": "network.openstack.org", "controllerKind": "DNSMasq", "DNSMasq": {"name":"bbc8d956-06c9-47f4-bd0e-f14a68c65863","namespace":"40218426-5c50-4f36-9a1b-58583d72ee5f"}, "namespace": "40218426-5c50-4f36-9a1b-58583d72ee5f", "name": "bbc8d956-06c9-47f4-bd0e-f14a68c65863", "reconcileID": "99c69042-60d2-4e7c-bb65-a40712780d3f", "req": "40218426-5c50-4f36-9a1b-58583d72ee5f/bbc8d956-06c9-47f4-bd0e-f14a68c65863"}
  2023-05-11T10:48:41.863+0200	INFO	Reconciling	{"controller": "service", "controllerGroup": "", "controllerKind": "Service", "Service": {"name":"some-service","namespace":"40218426-5c50-4f36-9a1b-58583d72ee5f"}, "namespace": "40218426-5c50-4f36-9a1b-58583d72ee5f", "name": "some-service", "reconcileID": "6dfa18ba-3ee2-4d3d-aba1-e36307f2434f", "req": "40218426-5c50-4f36-9a1b-58583d72ee5f/some-service"}
  2023-05-11T10:48:41.863+0200	INFO	controllers.Service	creating / updating DNSData	{"DNSHosts": []}
  2023-05-11T10:48:41.864+0200	INFO	controllers.Service	bbc8d956-06c9-47f4-bd0e-f14a68c65863-svc - operation: created
  2023-05-11T10:48:41.864+0200	INFO	Reconciling	{"controller": "service", "controllerGroup": "", "controllerKind": "Service", "Service": {"name":"some-service","namespace":"40218426-5c50-4f36-9a1b-58583d72ee5f"}, "namespace": "40218426-5c50-4f36-9a1b-58583d72ee5f", "name": "some-service", "reconcileID": "c9f96a24-d9db-4810-9c5a-9b9d8978a161", "req": "40218426-5c50-4f36-9a1b-58583d72ee5f/some-service"}
  2023-05-11T10:48:41.864+0200	INFO	controllers.Service	creating / updating DNSData	{"DNSHosts": [{"ip":"172.20.0.80","hostnames":["some-service.40218426-5c50-4f36-9a1b-58583d72ee5f.svc"]}]}
  2023-05-11T10:48:41.864+0200	INFO	Reconciling	{"controller": "dnsdata", "controllerGroup": "network.openstack.org", "controllerKind": "DNSData", "DNSData": {"name":"bbc8d956-06c9-47f4-bd0e-f14a68c65863-svc","namespace":"40218426-5c50-4f36-9a1b-58583d72ee5f"}, "namespace": "40218426-5c50-4f36-9a1b-58583d72ee5f", "name": "bbc8d956-06c9-47f4-bd0e-f14a68c65863-svc", "reconcileID": "3a59912a-5d7c-4b00-83ec-9107d5b75246", "req": "40218426-5c50-4f36-9a1b-58583d72ee5f/bbc8d956-06c9-47f4-bd0e-f14a68c65863-svc"}
  2023-05-11T10:48:41.864+0200	INFO	Reconciling	{"controller": "dnsmasq", "controllerGroup": "network.openstack.org", "controllerKind": "DNSMasq", "DNSMasq": {"name":"bbc8d956-06c9-47f4-bd0e-f14a68c65863","namespace":"40218426-5c50-4f36-9a1b-58583d72ee5f"}, "namespace": "40218426-5c50-4f36-9a1b-58583d72ee5f", "name": "bbc8d956-06c9-47f4-bd0e-f14a68c65863", "reconcileID": "65248cb1-8f27-487b-b7fb-945558437e9b", "req": "40218426-5c50-4f36-9a1b-58583d72ee5f/bbc8d956-06c9-47f4-bd0e-f14a68c65863"}
  2023-05-11T10:48:41.865+0200	INFO	---Test---	Getting DNSData	{"name": "40218426-5c50-4f36-9a1b-58583d72ee5f/bbc8d956-06c9-47f4-bd0e-f14a68c65863-svc"}
```
Failed sequence:
```
  2023-05-11T10:50:15.193+0200	INFO	---Test---	Creating	{"raw": {"apiVersion":"network.openstack.org/v1beta1","kind":"DNSMasq","metadata":{"name":"10a49d4e-96dc-4797-a38f-2c481b9e7dbc","namespace":"c1f18965-176a-49e7-b5b0-e367959794a8"},"spec":{"containerImage":"test-dnsmasq-container-image","dnsDataLabelSelectorValue":"dnsdata","externalEndpoints":[{"ipAddressPool":"ctlplane","loadBalancerIPs":["internal-lb-ip-1","internal-lb-ip-2"]}],"options":[{"key":"server","values":["1.1.1.1"]}]}}}
  2023-05-11T10:50:15.291+0200	INFO	Starting workers	{"controller": "service", "controllerGroup": "", "controllerKind": "Service", "worker count": 1}
  2023-05-11T10:50:15.291+0200	INFO	Reconciling	{"controller": "service", "controllerGroup": "", "controllerKind": "Service", "Service": {"name":"kubernetes","namespace":"default"}, "namespace": "default", "name": "kubernetes", "reconcileID": "a13ecc56-696e-4362-8521-877a992b4842", "req": "default/kubernetes"}
  2023-05-11T10:50:15.291+0200	INFO	Starting workers	{"controller": "dnsmasq", "controllerGroup": "network.openstack.org", "controllerKind": "DNSMasq", "worker count": 1}
  2023-05-11T10:50:15.291+0200	INFO	Starting workers	{"controller": "dnsdata", "controllerGroup": "network.openstack.org", "controllerKind": "DNSData", "worker count": 1}
  2023-05-11T10:50:17.213+0200	INFO	Reconciling	{"controller": "dnsmasq", "controllerGroup": "network.openstack.org", "controllerKind": "DNSMasq", "DNSMasq": {"name":"10a49d4e-96dc-4797-a38f-2c481b9e7dbc","namespace":"c1f18965-176a-49e7-b5b0-e367959794a8"}, "namespace": "c1f18965-176a-49e7-b5b0-e367959794a8", "name": "10a49d4e-96dc-4797-a38f-2c481b9e7dbc", "reconcileID": "be374dd3-a29d-4860-a98d-fd1003eb48bd", "req": "c1f18965-176a-49e7-b5b0-e367959794a8/10a49d4e-96dc-4797-a38f-2c481b9e7dbc"}
  2023-05-11T10:50:17.220+0200	INFO	Reconciling	{"controller": "service", "controllerGroup": "", "controllerKind": "Service", "Service": {"name":"some-service","namespace":"c1f18965-176a-49e7-b5b0-e367959794a8"}, "namespace": "c1f18965-176a-49e7-b5b0-e367959794a8", "name": "some-service", "reconcileID": "1b799208-e96e-4f79-84b9-408c58fb10a9", "req": "c1f18965-176a-49e7-b5b0-e367959794a8/some-service"}
  2023-05-11T10:50:17.220+0200	INFO	controllers.Service	creating / updating DNSData	{"DNSHosts": []}
  2023-05-11T10:50:17.223+0200	INFO	controllers.Service	10a49d4e-96dc-4797-a38f-2c481b9e7dbc-svc - operation: created
  2023-05-11T10:50:17.223+0200	INFO	Reconciling	{"controller": "dnsdata", "controllerGroup": "network.openstack.org", "controllerKind": "DNSData", "DNSData": {"name":"10a49d4e-96dc-4797-a38f-2c481b9e7dbc-svc","namespace":"c1f18965-176a-49e7-b5b0-e367959794a8"}, "namespace": "c1f18965-176a-49e7-b5b0-e367959794a8", "name": "10a49d4e-96dc-4797-a38f-2c481b9e7dbc-svc", "reconcileID": "1f64feda-542f-4552-9d15-b35aeb0c2404", "req": "c1f18965-176a-49e7-b5b0-e367959794a8/10a49d4e-96dc-4797-a38f-2c481b9e7dbc-svc"}
  2023-05-11T10:50:17.223+0200	INFO	Reconciling	{"controller": "service", "controllerGroup": "", "controllerKind": "Service", "Service": {"name":"some-service","namespace":"c1f18965-176a-49e7-b5b0-e367959794a8"}, "namespace": "c1f18965-176a-49e7-b5b0-e367959794a8", "name": "some-service", "reconcileID": "72c3caee-aa97-40f7-9b04-21d9fd3eb334", "req": "c1f18965-176a-49e7-b5b0-e367959794a8/some-service"}
  2023-05-11T10:50:17.223+0200	INFO	controllers.Service	creating / updating DNSData	{"DNSHosts": [{"ip":"172.20.0.80","hostnames":["some-service.c1f18965-176a-49e7-b5b0-e367959794a8.svc"]}]}
  2023-05-11T10:50:17.224+0200	INFO	Reconciling	{"controller": "dnsmasq", "controllerGroup": "network.openstack.org", "controllerKind": "DNSMasq", "DNSMasq": {"name":"10a49d4e-96dc-4797-a38f-2c481b9e7dbc","namespace":"c1f18965-176a-49e7-b5b0-e367959794a8"}, "namespace": "c1f18965-176a-49e7-b5b0-e367959794a8", "name": "10a49d4e-96dc-4797-a38f-2c481b9e7dbc", "reconcileID": "096dc5ed-66f8-4384-a577-b87b1c67ef21", "req": "c1f18965-176a-49e7-b5b0-e367959794a8/10a49d4e-96dc-4797-a38f-2c481b9e7dbc"}
  2023-05-11T10:50:17.224+0200	INFO	---Test---	Getting DNSData	{"name": "c1f18965-176a-49e7-b5b0-e367959794a8/10a49d4e-96dc-4797-a38f-2c481b9e7dbc-svc"}
  2023-05-11T10:50:17.225+0200	INFO	---Test---	Got DNSData	{"data": {"metadata":{"name":"10a49d4e-96dc-4797-a38f-2c481b9e7dbc-svc","namespace":"c1f18965-176a-49e7-b5b0-e367959794a8","uid":"d97868b7-ef1d-4efc-a482-ab02766a2e59","resourceVersion":"225","generation":1,"creationTimestamp":"2023-05-11T08:50:17Z","ownerReferences":[{"apiVersion":"network.openstack.org/v1beta1","kind":"DNSMasq","name":"10a49d4e-96dc-4797-a38f-2c481b9e7dbc","uid":"d0e6aa0a-29cc-4d84-8d9b-8a4d7bb9da9c","controller":true,"blockOwnerDeletion":true}],"finalizers":["DNSData"],"managedFields":[{"manager":"functional.test","operation":"Update","apiVersion":"network.openstack.org/v1beta1","time":"2023-05-11T08:50:17Z","fieldsType":"FieldsV1","fieldsV1":{"f:metadata":{"f:finalizers":{".":{},"v:\"DNSData\"":{}},"f:ownerReferences":{".":{},"k:{\"uid\":\"d0e6aa0a-29cc-4d84-8d9b-8a4d7bb9da9c\"}":{}}},"f:spec":{".":{},"f:dnsDataLabelSelectorValue":{}}}}]},"spec":{"dnsDataLabelSelectorValue":"dnsdata"},"status":{}}}
  [FAILED] in [It] - /home/gibi/upstream/git/openstack-k8s-operators/infra-operator/test/functional/service_controller_test.go:81 @ 05/11/23 10:50:17.225
```

Reply:
So the DNSData goes through a two step reconciliation. First it is created with empty DNSHost then it is updated with a proper DNSHost. I did not dig deeper so I'm not sure it is bug in the controller or it is a normal behavior. But it explains that why the test fails sometimes. If the test reads the DNSData after it was created but before it was updated then it will see an empty DNSHost list. While if the test reads it after both create and update then it gets the populated list.


* a) If this is not the expected behavior of the controller then the controller should be fixed.
* b) If this is the expected behavior of the controller then you can wrap the L78-L83 into an Eventually block to *wait* for the DNSData to be fully populated.
  ```golang
			Eventually(func(g Gomega) {
				dnsdata := GetDNSData(svcDNSData)
				g.Expect(dnsdata).To(Not(BeNil()))
				g.Expect(dnsdata.Spec.DNSDataLabelSelectorValue).To(Equal("dnsdata"))
				g.Expect(dnsdata.Spec.Hosts).To(Not(BeNil()))
				g.Expect(dnsdata.Spec.Hosts[0].IP).To(Equal("172.20.0.80"))
				g.Expect(dnsdata.Spec.Hosts[0].Hostnames[0]).To(Equal(fmt.Sprintf("some-service.%s.svc", namespace)))
			}, timeout, interval).Should(Succeed())
```

Reply:
Thanks for helping me on this. Not sure why I have not considered using Eventually here.With the current implementation its valid to have empty DNSHost. This is to make sure that when all services with annotations got removed we remove all entries from the DNSData CR. I'll update the PR with using Eventually